spec/src/main/asciidoc/microprofile-lra-spec.adoc

```diff
@@ -322,6 +326,35 @@ annotations defined in this specification then the LRA will not be made
 available to the invoked method. However, any outgoing JAX-RS invocations
 made during the execution of the method will still carry the original context.
 
+The user is allowed to manually assign the `context header` on JAX-RS requests
+and responses:
+
+===== Setting the context on outgoing JAX-RS requests
+
+If the context is set on an outgoing JAX-RS request then it becomes the `active
+context` for that invocation. When the invocation returns the `active context`
+will depend on the behaviour of the remote resource:
+
+- if the remote resouce ends the context (and does not start a new one) then
+  the `active context` will be whatever the client was using before it manually set
+  the context on the outgoing header;
+- if the remote resouce does not end the context (and does not start a new one)
```

Comment:
In the following scenario

@LRA
Response doSomething() {
  // here we have LRAid = 123
  client.  ... .header(LRA.LRA_HTTP_HEADER, "456").get() ;   // We set LRAid 456 to outgoing call (I know they are URI, just for simplicity)
 // Does this means the 'Active Context' is now LRAid = 456 and no longer 123?
  client.  ... .get();  .// Without specifying the header, does this mean we are using LRAid 456?

-> If so, this will be confusing. We specify the header for a certain REST call but it should not affect the 'Current Context' of method doSomething().
And also, how will the system know that the called endpoint ended the LRA (as mentioned in previous bullet) because it should not alter the 'current Context' of method doSomething()? 
Is the LRA_HTTP_HEADER only added to responses when the LRA is still active (end = false)?

Reply:
The code in the doSomething() method has explicitly set the context on the outgoing request so it is managing the context manually - ie the business developer understands the resulting behaviour since he has read this bit of the spec. Now if the service invoked by the doSomething() method does not end the LRA and if the code in the doSomething() method wants the next request to use the original context then he just needs to explicitly set it on any other outgoing invocations made inside the doSomething() method.

Note that I do say in the following paragraph that

> The user MAY manage the context manually by tracking in progress LRAs (via the context header) and ensure he manually sets the desired context on outgoing JAX-RS invocations.

To your second point, I think the spec is clear (but we can clarify if not) that if the LRA annotation does not automatically end an LRA it becomes the active one for the caller (otherwise the LRA id would be lost/unavailable to the caller).

To your final point, yes, the LRA_HTTP_HEADER is only added to responses when the LRA is still active (end = false). This is because the `context header` is reserved for the `active context`.

Reply:
I think the problem is that I am wanting to support "implicit context propagation" where the spec implementation manages the association. We currently have both explicit (where the user is allowed to manage the context) and implicit. I acknowledge that supporting both can be a challenge to specify in a clear and unambiguous way. Manage thread association can also be tricky in reactive environments.

So would people be more comfortable if we say that the user has to manage the context on outgoing requests (I think Martin asked for this a while back) and the container manages the response header according to what is dictated by the @LRA annotation.

If so then I will update this PR to remove the requirement for "implicit context propagation". Thoughts?

Reply:
Implicit context propagation is good in my opinion and probably 99% of the cases.
But when the developer sets the context, it should only be for that single request which doesn't affect the current context.

I think this can easily be achieved by not looking at the value on the Response. (Which also covers then the case where the called endpoint starts a new; completely separate LRA)

We can discus this on next meeting

Reply:
+1 (we also got +1's from everyone present at today's hangout).
@rdebusscher Will you check that my last commit (https://github.com/eclipse/microprofile-lra/pull/124/commits/4faab7d2f84f135bd19896a8117c7de730ceb02f) matches your expectation. Thanks